### PR TITLE
Update k8s tests and libraries to v1.27.2

### PR DESCRIPTION
**Pull request by nathanjsweet**
Updating k8s libs to v1.27.0 requires updating sigs.k8s.io/controller-runtime to an unreleased version with https://github.com/kubernetes-sigs/controller-runtime/pull/2223 in place. This in turn requires some additional adjustments to changed APIs in the controller-runtime, mainly in operator/pkg/gateway-api.

The service.kubernetes.io/topology-aware-hints annotation got deprecated with k8s 1.27 in favor of service.kubernetes.io/topology-mode (see https://github.com/kubernetes/kubernetes/pull/116522). This change keeps support for the deprecated annotation. Support for the new annotation will be added in a successive commit.

```release-note
Add support for Kubernetes v1.27
```

### CODE REVIEW

**Review by zacharysarah — APPROVED** [review]
LGTM from a docs perspective.

**Comment by pchaigno** [comment]
/test-1.27-net-next

File: contrib/vagrant/scripts/helpers.bash [hunk 1]
```diff
@@ -67,8 +67,8 @@ cluster_dns_ipv6=${K8S_CLUSTER_DNS_IPV6:-"FD03::A"}
 cluster_api_server_ipv4=${K8S_CLUSTER_API_SERVER_IPV4:-"172.20.0.1"}
 cluster_api_server_ipv6=${K8S_CLUSTER_API_SERVER_IPV6:-"FD03::1"}
 
-k8s_version="v1.26.0-rc.0"
-etcd_version="v3.5.5"
+k8s_version="v1.27.0"
+etcd_version="v3.5.7"
```
Comment (bimmlerd) [inline contrib/vagrant/scripts/helpers.bash:71]: I remember we had issues with this etcd version - https://github.com/cilium/cilium/issues/23760 something about the image being distroless etc- was that fixed in the meantime?

**Review by bimmlerd — COMMENTED** [review]
otherwise looks good from contributing's side

**Comment by nbusseneau** [comment]
Closed and re-opened to force re-trigger CI checks, as image building failed and everything else failed subsequently.

**Comment by nbusseneau** [comment]
Please note that we had to revert the previous PR at #24837 due to errors in the upstream tests, so on top of `/test-1.27-net-next` we should also iterate on `/test-upstream-k8s`.

**Comment by odinuge** [comment]
👋 

Is there a timeline for when these libs will be bumped? It would be nice to get some of the bugfixes in upstream, but right now the `main` branch is stuck on `v0.26.0` even though `v0.26.5` has been released. Is there any chance we can do a patch bump first; then go for `v0.27.1`/`v0.27.2`? 😄

**Comment by nathanjsweet** [comment]
> Is there a timeline for when these libs will be bumped? It would be nice to get some of the bugfixes in upstream, but right now the main branch is stuck on v0.26.0 even though v0.26.5 has been released. Is there any chance we can do a patch bump first; then go for v0.27.1/v0.27.2?

Unfortunately not, there are some breaking interface changes between some of the K8s libraries and ours. This PR is my top priority though. So very soon.

**Comment by nathanjsweet** [comment]
I'll try to leave daily updates (excluding next Monday, which is a US Holiday) going forward.

I made good progress on this PR, unfortunately upgrading the `controller-runtime` package (which we need to do for [this PR](https://github.com/kubernetes-sigs/controller-runtime/pull/2223)) from 0.14.1 to 0.15.0 created all kinds of breakages in the tests for our `operator/pkg/gateway-api` package. Hopefully, this will be fixed early next week (maybe even Monday), and we can proceed.

File: contrib/vagrant/scripts/helpers.bash (hunk 1 above)
Comment (nathanjsweet) [reply contrib/vagrant/scripts/helpers.bash:71]: This is just for testing. It doesn't actually use the docker image either. This version slug is used to download etcd and runs it as a systemd process.

**Comment by nathanjsweet** [comment]
/test-1.27-net-next

**Review by meyskens — APPROVED** [review]
Code changes to Gateway API look good! Thanks!

**Review by joamaki — DISMISSED** [review]
Please also update `test/controlplane/k8s_versions.txt` and regenerate the tests (see `test/controlplane/README.md`).
EDIT: Totally fine with separate PR for this too if you prefer that route! The changes otherwise look good here.

**Review by bimmlerd — APPROVED** [review]

**Review by rolinh — APPROVED** [review]
Vendor changes lgtm

**Review by christarazi — APPROVED** [review]

**Review by rastislavs — APPROVED** [review]
LGTM for the BGP part

**Review by brlbil — APPROVED** [review]

**Comment by mathpl** [comment]
@nathanjsweet would like https://github.com/cilium/cilium/pull/25913 in order to understand the issue(s) here. :)

**Comment by christarazi** [comment]
/test-1.26-net-next

**Pull request merged**